chain/chain/src/chain.rs

```diff
@@ -5099,21 +5101,36 @@ impl<'a> ChainUpdate<'a> {
                     .get_split_shard_uids(shard_uid.shard_id())
                     .unwrap_or_else(|| panic!("invalid shard layout {:?}", next_epoch_shard_layout))
                     .len() as NumShards;
+
                 let total_gas_used = chunk_extra.gas_used();
                 let total_balance_burnt = chunk_extra.balance_burnt();
-                let gas_res = total_gas_used % num_split_shards;
+
+                let mut gas_res = total_gas_used % num_split_shards;
                 let gas_split = total_gas_used / num_split_shards;
-                let balance_res = (total_balance_burnt % num_split_shards as u128) as NumShards;
+
+                let mut balance_res = (total_balance_burnt % num_split_shards as u128) as NumShards;
                 let balance_split = total_balance_burnt / (num_split_shards as u128);
                 let gas_limit = chunk_extra.gas_limit();
                 let outcome_root = *chunk_extra.outcome_root();
 
                 let mut sum_gas_used = 0;
                 let mut sum_balance_burnt = 0;
+
+                // The gas and balance distribution assumes that we have a result for every split shard.
+                // TODO(resharding) make sure that is the case.
+                assert_eq!(num_split_shards, results.len() as u64);
```

Comment:
Hmm... I wonder if an assert is best here or if we should use `tracing::error` here instead?  Will something else fail further down if this assert does not hold?